### PR TITLE
Some preliminary work for nnet3 support of convnets

### PR DESCRIPTION
**Pull request by danpovey**
This may be of interest to @galv.
I may ask you for help at some point, either to finish the code or to help compare with CuDNN.
I was inspired to do this after looking at your PR.  My hope is that for the normal fairly small-sized e.g. 3x3 convolution kernels, it will be unnecessary to use CuDNN in order to get optimal speed.  Read convolution.h and you'll see why.

### CODE REVIEW

**Comment by galv** [comment]
@danpovey I've seen this in my inbox. I haven't looked at this in detail yet, though.

File: src/nnet3/nnet-common.cc [hunk 1]
```diff
@@ -535,7 +535,8 @@ void PrintIntegerVector(std::ostream &os,
   os << "]";
 }
 
-
+// this will be the most negative number representable as int32.
+const int kNoTime = std::numeric_limits<BaseFloat>::min();
```
Comment (galv) [inline src/nnet3/nnet-common.cc]: Shouldn't BaseFloat be int32 here?

**Comment by danpovey** [comment]
yes, thanks..

File: src/nnet3/nnet-convolutional-component.h [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+  DistributeComponent(int32 input_dim, int32 output_dim) {
+    Init(input_dim, output_dim);
+  }
+  DistributeComponent(): input_dim_(0), output_dim_(0) { }
+  virtual int32 InputDim() const { return input_dim_; }
+  virtual int32 OutputDim() const { return output_dim_; }
+
+  // use the default Info() function.
+  virtual void InitFromConfig(ConfigLine *cfl);
+  virtual std::string Type() const { return "DistributeComponent"; }
```
Comment (galv) [inline src/nnet3/nnet-convolutional-component.h]: Should this really return "DistributeComponent"?

**Comment by danpovey** [comment]
that was just a draft, I'll work on that later.

**Comment by danpovey** [comment]
Closing this PR, combined it with batch-norm.

**Pull request closed without merging**